File: android_gemini/core/common/src/main/java/com/zarabandajose/transcripto/core/common/Envelope.kt
package com.zarabandajose.transcripto.core.common

import android.util.Base64

object Envelope {
    fun pack(method: String, salt: String?, payload: String): String {
        val saltBase64 = salt?.let { Base64.encodeToString(it.toByteArray(), Base64.NO_WRAP) } ?: ""
        return "method=$method;salt=$saltBase64;payload=$payload"
    }

    fun unpack(envelope: String): Triple<String, String?, String>? {
        if (!envelope.startsWith("method=")) return null
        val parts = envelope.split(";")
        if (parts.size != 3) return null
        val method = parts[0].substringAfter("=")
        val saltBase64 = parts[1].substringAfter("=")
        val payload = parts[2].substringAfter("=")
        val salt = if (saltBase64.isNotEmpty()) {
            String(Base64.decode(saltBase64, Base64.NO_WRAP))
        } else {
            null
        }
        return Triple(method, salt, payload)
    }
}

File: android_flutter/transcripto/packages/domain/lib/src/envelope.dart
class Envelope {
  final String method;
  final String saltBase64; // may be empty
  final String payload;
  const Envelope({required this.method, required this.saltBase64, required this.payload});

  @override
  String toString() => 'method=$method;salt=$saltBase64;payload=$payload';

  static Envelope? tryParse(String input) {
    // Expected: method=...;salt=...;payload=...
    if (!input.startsWith('method=')) return null;
    final parts = input.split(';');
    if (parts.length < 3) return null;
    final map = <String, String>{};
    for (final p in parts) {
      final idx = p.indexOf('=');
      if (idx <= 0) continue;
      final k = p.substring(0, idx);
      final v = p.substring(idx + 1);
      map[k] = v;
    }
    if (!map.containsKey('method') || !map.containsKey('salt') || !map.containsKey('payload')) {
      return null;
    }
    return Envelope(method: map['method']!, saltBase64: map['salt']!, payload: map['payload']!);
  }
}
